### ISSUE TITLE
TestThreadAuthentication.test_curve_user_id appears to be racy

### ISSUE THREAD

**Issue by mwhudson** [zeromq/pyzmq#1337]
This is a frustrating issue as it only appears on Ubuntu's CI infrastructure and not at all on my machine but I do have a theory as to what is happening. The symptom is that TestThreadAuthentication.test_curve_user_id fails like this:

```
=================================== FAILURES ===================================
_________________ TestThreadAuthentication.test_curve_user_id __________________

self = <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>

    @skip_pypy
    def test_curve_user_id(self):
        """threaded auth - CURVE"""
        self.auth.allow('127.0.0.1')
        certs = self.load_certs(self.secret_keys_dir)
        server_public, server_secret, client_public, client_secret = certs
    
        self.auth.configure_curve(domain='*', location=self.public_keys_dir)
        server = self.socket(zmq.PULL)
        server.curve_publickey = server_public
        server.curve_secretkey = server_secret
        server.curve_server = True
        client = self.socket(zmq.PUSH)
        client.curve_publickey = client_public
        client.curve_secretkey = client_secret
        client.curve_serverkey = server_public
>       assert self.can_connect(client, server)
E       AssertionError: assert False
E        +  where False = <bound method TestThreadAuthentication.can_connect of <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>>(<zmq.sugar.socket.Socket object at 0x7fab2a042850>, <zmq.sugar.socket.Socket object at 0x7fab2a042650>)
E        +    where <bound method TestThreadAuthentication.can_connect of <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>> = <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>.can_connect

/usr/lib/python3/dist-packages/zmq/tests/test_auth.py:327: AssertionError
```

(full log here https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-focal/focal/amd64/p/pyzmq/20191101_193354_9ea5a@/log.gz and many more examples linked from http://autopkgtest.ubuntu.com/packages/p/pyzmq/focal/amd64).

What I *think* is happening is that the can_connect call is happening before the configuration of the authentication method has actually completed. The log output for the above failure is:

```
------------------------------ Captured log call -------------------------------
base.py                     53 DEBUG    Starting
thread.py                   86 DEBUG    auth received API command b'ALLOW'
base.py                     73 DEBUG    Allowing 127.0.0.1
base.py                    207 DEBUG    version: b'1.0', request_id: b'1', domain: '', address: '127.0.0.1', identity: b'', mechanism: b'CURVE'
base.py                    222 DEBUG    PASSED (whitelist) address=127.0.0.1
base.py                    375 DEBUG    ZAP reply code=b'400' text=b'Unknown domain'
thread.py                   86 DEBUG    auth received API command b'CURVE'
base.py                    114 DEBUG    Configure curve: *[/tmp/tmpewmxhtfc/public_keys]
base.py                    207 DEBUG    version: b'1.0', request_id: b'1', domain: '', address: '127.0.0.1', identity: b'', mechanism: b'CURVE'
base.py                    222 DEBUG    PASSED (whitelist) address=127.0.0.1
base.py                    356 DEBUG    ALLOWED (CURVE) domain=* client_key=b'iHzx::[5<yCgiHW&3&St(n3jRJ}pg&6ov.TOH7s^'
base.py                    375 DEBUG    ZAP reply code=b'200' text=b'OK'
thread.py                   86 DEBUG    auth received API command b'TERMINATE'
```

You can see that a connection is rejected with "Unknown domain" before the CURVE api method message is received, and locally in passing tests I don't see that. There is another request later which succeeds so maybe I'm failing to understand how all this works and I'm barking up the wrong tree but in any case, the test is definitely failing...

**Comment by mwhudson** [comment]
I managed to make this fail locally by inserting a "time.sleep(0.01)" into AuthenticationThread.run just before the "while True" loop. I guess ThreadAuthenticator.start should not return until the AuthenticationThread has reached this point. Not sure how to ensure that.

**Comment by mwhudson** [comment]
OH, I see what's happening here (I think, given that I've never looked into pyzmq before a day or so ago): the issue is that the loop in AuthenticationThread.run only handles at most one "api" request before it processes "zap" requests. So here's what's happening in the failing test:

The test calls self.auth.allow (which sends an api request) and self.auth.configure_curve (which sends another api request) and then can_cancel calls client.poll which ends up sending a zap request.

If we lose the race, all of this happens before AuthenticationThread.run starts processing events. It checks the "api" pipe first and processes the "ALLOW" message. Then it checks the "zap" pipe which finds a connection request. Because the CURVE api message has not been processed, this request is rejected. Next time around the loop, the CURVE api message is processed, but it's too late and the test fails.

I don't know enough about pyzmq / 0mq to know how best to fix this in full generality. I guess AuthenticationThread._handle_pipe could make non-blocking reads from the pipe until there are no more messages?

**Comment by mwhudson** [comment]
This patch seems to make the test robust against quite large delays inserted anywhere I can think of:

```diff
--- ./zmq/auth/thread.py	2019-11-04 11:48:22.136433688 +1300
+++ .pybuild/cpython3_3.7/build/zmq/auth/thread.py	2019-11-04 12:45:02.386092048 +1300
@@ -51,7 +51,18 @@
                 break  # interrupted
 
             if self.pipe in socks and socks[self.pipe] == zmq.POLLIN:
-                terminate = self._handle_pipe()
+                # Make sure all API requests are processed before
+                # looking at the ZAP socket.
+                while True:
+                    try:
+                        msg = self.pipe.recv_multipart(flags=zmq.NOBLOCK)
+                    except zmq.Again:
+                        self.log.debug("zmq.Again")
+                        break
+
+                    terminate = self._handle_pipe(msg)
+                    if terminate:
+                        break
                 if terminate:
                     break
 
@@ -69,15 +80,12 @@
         if not msg: return
         self.authenticator.handle_zap_message(msg)
 
-    def _handle_pipe(self):
+    def _handle_pipe(self, msg):
         """
         Handle a message from front-end API.
         """
         terminate = False
 
-        # Get the whole message off the pipe in one go
-        msg = self.pipe.recv_multipart()
-
         if msg is None:
             terminate = True
             return terminate
```

**Comment by minrk** [comment]
@mwhudson thanks for digging in! That patch looks fine to me, would you like to make a PR? Or I can apply the patch myself if you'd prefer.

**Comment by mwhudson** [comment]
Sure, done. I think there is a chance this might fix #953 too but I'm not sure.
